From 39812bd02da37d2e73b36bcbd998d40e1d48de0f Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Fri, 7 Sep 2012 14:20:44 +0300
Subject: [PATCH 364/370] conf: fix memory leak in test_obj_fill()

+ c2_conf_reg_fini(): set ->co_status of deleted objects to C2_CO_MISSING,
  because _generic_invariant() expects unmounted objects to be C2_CO_MISSING.

+ Improve readability of _generic_obj_invariant().
---
 conf/obj_ops.c | 5 ++---
 conf/reg.c     | 7 ++++++-
 conf/ut/obj.c  | 2 ++
 3 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/conf/obj_ops.c b/conf/obj_ops.c
index c222b1e..50b6002 100644
--- a/conf/obj_ops.c
+++ b/conf/obj_ops.c
@@ -69,9 +69,8 @@ static bool _generic_obj_invariant(const void *bob)
 		buf_is_valid(&obj->co_id) && obj->co_ops != NULL &&
 		C2_IN(obj->co_status,
 		      (C2_CS_MISSING, C2_CS_LOADING, C2_CS_READY)) &&
-		(obj->co_status == C2_CS_READY ? obj->co_mounted :
-		 obj->co_nrefs == 0) &&
-		ergo(!obj->co_mounted, obj->co_status == C2_CS_MISSING);
+		ergo(!obj->co_mounted, obj->co_status == C2_CS_MISSING) &&
+		ergo(obj_is_stub(obj), obj->co_nrefs == 0);
 }
 
 struct c2_conf_obj *c2_conf__dir_create(void);
diff --git a/conf/reg.c b/conf/reg.c
index 9eb1155..42b219d 100644
--- a/conf/reg.c
+++ b/conf/reg.c
@@ -79,8 +79,13 @@ void c2_conf_reg_fini(struct c2_conf_reg *reg)
 
 	c2_tl_for(c2_conf_reg, &reg->r_objs, obj) {
 		c2_conf_reg_tlist_del(obj);
-		/* don't let concrete invariants check relations */
+
+		/* Don't let concrete invariants check relations. */
 		obj->co_mounted = false;
+		/* _generic_invariant() expects unmounted objects to
+		 * have C2_CS_MISSING status. */
+		obj->co_status = C2_CS_MISSING;
+
 		c2_conf_obj_delete(obj);
 	} c2_tl_endfor;
 
diff --git a/conf/ut/obj.c b/conf/ut/obj.c
index 58a73b4..77eb125 100644
--- a/conf/ut/obj.c
+++ b/conf/ut/obj.c
@@ -93,6 +93,8 @@ void test_obj_fill(void)
 			rc = c2_conf_obj_fill(obj, &xobjs[i], &reg);
 			C2_UT_ASSERT(rc == 0);
 		}
+
+		c2_conf_reg_add(&reg, obj);
 	}
 
 	c2_confx_fini(xobjs, nr_objs);
-- 
1.8.3.2

